File: solutions/python/robot-name/1/robot_name.py
import random


class Robot:
    _names = set()

    def __init__(self):
        self._name = self._generate_random_unique_name()

    def _generate_random_unique_name(self):
        def randl(): return chr(random.randint(ord('A'), ord('Z')))
        def randn(): return str(random.randint(1, 9))

        name = ''.join(fn() for fn in [randl, randl, randn, randn, randn])
        if name in Robot._names:
            return self._generate_random_unique_name()

        Robot._names.add(name)
        return name

    def reset(self):
        self._name = self._generate_random_unique_name()
        Robot._names.remove(self._name)

    @property
    def name(self) -> str:
        return self._name

File: solutions/ruby/robot-name/1/robot_name.rb
class Robot
  ALPHABET = ('a'..'z').to_a.freeze
  @@names = []

  attr_reader :name

  def self.forget
    @@names = []
  end

  def initialize
    reset
  end

  def reset
    @name = new_name
  end

  private
  def new_name
    name = next_possible_name
    @@names << name
    name
  end

  def next_possible_name
    return 'AA000' if @@names.length == 0

    a,b,c,d,e = @@names.last.chars
    e = e.to_i + 1
    if e == 10
      e = 0
      d = d.to_i + 1
    end
    if d == 10
      d = 0
      c = c.to_i + 1
    end
    if c == 10
      c = 0
      b = (b.ord + 1).chr
    end
    if b > 'Z'
      b = 'A'
      a = (a.ord + 1).chr
    end

    [a,b,c,d,e].join
  end
end